### PR TITLE
MAINT: Refactor Fit / Zoom parameters

### PR DESCRIPTION
**Pull request by MartinThoma**
Introduce a new `PyPDF2.generic.Fit` class which captures the type and parameter for how a page should be fit in the viewer (e.g. when clicking on a PDF-internal link).

The class has one method for each fit type which allows users to discover the different types via their IDE, e.g. `Fit.xyz(left=123, top=456, zoom=2)`.

## Breaking Change

This PR **introduces a breaking change** that needs a major version bump. Affected are the following methods:

* `PdfMerger.add_outline_item` used `fit: FitType` and `*args: ZoomArgType` parameters
* `PdfWriter.add_outline_item` used `fit: FitType` and `*args: ZoomArgType` parameters
* `AnnotationBuilder.link`  used `fit: FitType = "/Fit"` and `fit_args: Tuple[ZoomArgType, ...] = tuple()` instead.

Instead of having two arguments, we now have only one. To that argument, an object of the new `Fit` class must be passed.

## Why *args is problematic

Using the `*args` pattern is problematic for two reasons:

* **User-code readability**:
  * People cannot use the more expressive keyword-argument syntax in their code for the non-fit parameters
  * People may or may not use the keyword-parameter for the fit parameters
* **Library extensions**: PyPDF2 cannot easily add new parameter to those functions; #1371 is the latest PR that stumbled over this issue
* **Two parameter for one thing**: In general I don't like if we have two parameters for one topic

### CODE REVIEW

**Comment by MartinThoma** [comment]
@pubpub-zz I would introduce this breaking change before #1371. Then you don't need a breaking change in that PR anymore, right?

**Comment by MartinThoma** [comment]
@pubpub-zz @MasterOdin What do you think about this proposal?

File: PyPDF2/_merger.py [hunk 1]
```diff
@@ -669,7 +675,13 @@ def addBookmark(
         """
         deprecate_with_replacement("addBookmark", "add_outline_item")
         return self.add_outline_item(
-            title, pagenum, parent, color, bold, italic, fit, *args
+            title,
+            pagenum,
+            parent,
```
Comment (pubpub-zz) [inline PyPDF2/_merger.py:680]: If we change the arguments, what about adding before parameter at the same time ?

File: PyPDF2/_merger.py [hunk 2]
```diff
@@ -54,9 +54,11 @@
 from .constants import PagesAttributes as PA
 from .constants import TypArguments, TypFitArguments
 from .generic import (
+    DEFAULT_FIT,
```
Comment (pubpub-zz) [inline PyPDF2/_merger.py:57]: Instead of DEFAULT_FIT, I would say PAGE_FIT

File: PyPDF2/generic/_fit.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+class Fit:
+    def __init__(
+        self, fit_type: str, fit_args: Tuple[Union[None, float, Any], ...] = tuple()
+    ):
+        from ._base import NameObject, NullObject, NumberObject
+
+        self.fit_type = NameObject(fit_type)
+        self.fit_args = [
+            NullObject() if a is None or isinstance(a, NullObject) else NumberObject(a)
```
Comment (pubpub-zz) [inline PyPDF2/generic/_fit.py:12]: I would say that NumberObject should be replaced  by FloatObject

File: PyPDF2/_merger.py (hunk 1 above)
Comment (MartinThoma) [reply PyPDF2/_merger.py:680]: In the same release, but not the same pr :-)

**Comment by MasterOdin** [comment]
Is there value in allowing `fit` to be specified as a string for the types that don't require any additional arguments (e.g. `/Fit`)? Not sure how prevalent usage was for setting this to know how much of a breaking change this is.

If this breaking change does happen, does that mean all of the various deprecated functions and things will be removed as well?

File: PyPDF2/generic/_fit.py (hunk 3 above)
Comment (MartinThoma) [reply PyPDF2/generic/_fit.py]: Thank you - done :-)

File: PyPDF2/_merger.py (hunk 2 above)
Comment (MartinThoma) [reply PyPDF2/_merger.py]: Done!

**Pull request merged**